### DISCUSSION TITLE
Issue with "-virtual-pixel None" and masked "-composite"

### DISCUSSION THREAD

**Discussion by GeeMack** [ImageMagick/ImageMagick#5609] [Development]
Using ImageMagick 7.1.0-49 Q16-HDRI x64 on Windows 10.

I've encountered an issue when doing a simple composite using two images and a mask. The problem occurs when "-virtual-pixel" is set to "None" or "Transparent".

This command works producing a red-to-blue gradient as expected...

    magick -size 240x240 xc:blue xc:red gradient: -composite test_right.png

![test_right](https://user-images.githubusercontent.com/65475279/193386693-d667a0a0-2080-41a0-86e3-c220b3f4edb9.png)

The next command with `-virtual-pixel none` appears to ignore the mask and produces an all red result...

    magick -size 240x240 -virtual-pixel none xc:blue xc:red gradient: -composite test_wrong.png

![test_wrong](https://user-images.githubusercontent.com/65475279/193386698-4848fc0a-0f78-4c6e-a203-d29dafd9c351.png)

And this command with `-virtual-pixel mirror` produces a gradient result as expected...

    magick -size 240x240 xc:blue xc:red gradient: -virtual-pixel mirror -composite test_right2.png

![test_right2](https://user-images.githubusercontent.com/65475279/193386702-11709943-0607-4d4d-a717-8e947d81edbb.png)

With all other "virtual-pixel" settings the alpha composite seems to work as expected. When it's set to "None" or "Transparent" the masked composite fails to work.

**Comment by fmw42** [comment]
Does it work if you move -virtual-pixel none to after gradient:.  You have split -size and xc: with something in-between. Not sure I would do that.

**Reply by GeeMack** [reply]
From what I can tell if the "-virtual-pixel none" comes before the "-composite" it fails.

**Comment by snibgo** [comment]
I confirm the problem with IM v7.1.0-49 on Windows 8.1.
```
magick -size 240x240 xc:blue xc:red gradient: -virtual-pixel none -compose Over -composite x.png
```
The result is entirely red. I see no reason why "`-virtual-pixel`" should affect the result.

EDIT to add:

The problem also occurs (of course) with "`-virtual-pixel transparent`".

The problem occurred in v7.0.10-45, so this isn't a new bug.

The problem did not occur in v6.9.9-50.

This seems to be a bug, possibly from the start of version 7.

**Comment by GeeMack** [comment]
Additional comment about this issue...

The mask composite will fail if the `-virtual-pixel` is set to "None" or "Transparent" somewhere before the composite, *even if it is changed to another setting before the composite*. This command...

    magick -size 240x240 xc:blue xc:red gradient: -virtual-pixel none -virtual-pixel edge -composite test_fail.png

... should output a red-to-blue gradient, but only produces an all-red image.

**Comment by snibgo** [comment]
> ... even if it is changed to another setting before the composite.

Confirmed with IM v7.1.0-49 on Windows 8.1.

```
magick -size 240x240 xc:blue xc:red gradient: -virtual-pixel none -virtual-pixel Edge -compose Over -composite x.png
```
The result is red (which is wrong).

However, if we "-alpha off" after the two "-virtual-pixel" operations:
```
magick -size 240x240 xc:blue xc:red gradient: -virtual-pixel none -virtual-pixel Edge -alpha off -compose Over -composite x.png
```
... the result is a correct red-blue gradient.

We might think that "-alpha off" would fix the problem with just one "-virtual-pixel". But it doesn't.
```
magick -size 240x240 xc:blue xc:red gradient: -virtual-pixel none -alpha off -compose Over -composite x.png
```
The result is, again, just red.
